pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2018-2019 Authors of Cilium
+
+package ipcache
+
+import (
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var (
+	// idMDMU protects the IdentityMetadata map.
+	//
+	// If this mutex will be held at the same time as the IPCache mutex,
+	// this mutex must be taken first and then take the IPCache mutex in
+	// order to prevent deadlocks.
+	idMDMU lock.RWMutex
+	// IdentityMetadata maps IP prefixes (x.x.x.x/32) to their labels.
+	IdentityMetadata = make(map[string]labels.Labels)
+)
+
+// UpsertMetadata upserts a given IP and its corresponding labels associated
+// with it into the IdentityMetadata map. The given labels are not modified nor
+// is its reference saved, as they're copied when inserting into the map.
+func UpsertMetadata(prefix string, lbls labels.Labels) {
+	l := labels.NewLabelsFromModel(nil)
+	l.MergeLabels(lbls)
+
+	idMDMU.Lock()
+	if cur, ok := IdentityMetadata[prefix]; ok {
+		l.MergeLabels(cur)
```

Comment:
Long-term, the idea will be to cover all identities that are associated with an IP that is covered by the prefix. So when we allocate local/CIDR identities, we look up the IdentityMetadata map and then add all of those labels into the set used for allocation. This way we should be able to associate ~arbitrary labels with IPs (and prefixes!).

To support that fully, I think we'll need for the `IdentityMetadata` structure to support multiple mappings for each prefix, then these will be combined at identity allocation time. Furthermore the lookup to find all of the labels that should be associated with a particular IP likely needs to be some sort of LPM. There's probably a bunch more nuance as we look into the details of such an implementation.

However, for now there's only one user-facing mechanism to associate the labels (for the labels to be `reserved:kube-apiserver`), and the association can only occur for specific IPs. Internally this PR is also handing the `reserved:remote-node` label through this code as well. So in practice it's fine to use a hashmap here and do the merging of labels during upsert. In practice this means that in the average case with this PR, this only affects one identity (the identity for the IP where the apiserver lives). Technically it could also affect other apiserver IPs if you have multiple of them and each of them is outside the cluster (and hence uses unique CIDR identities for each IP).

Reply:
I've added a comment to imply this; the commit "ipcache, policy: Inject labels from identity metadata" also describes it to a degree (probably with some details that we should import directly into the code so it's easier to reference).

Reply:
(I included the diagram in the code in the latest version since that seemed like the obvious thing to also include).